SpatialGDK/Source/SpatialGDK/Private/EngineClasses/Components/SpatialPingComponent.cpp

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "EngineClasses/Components/SpatialPingComponent.h"
+
+#include "GameFramework/PlayerController.h"
+#include "Net/UnrealNetwork.h"
+#include "TimerManager.h"
+#include "Engine/World.h"
+
+USpatialPingComponent::USpatialPingComponent(const FObjectInitializer& ObjectInitializer) : Super(ObjectInitializer)
+{
+	PrimaryComponentTick.bCanEverTick = false;
+	PrimaryComponentTick.bStartWithTickEnabled = false;
+
+	bReplicates = true;
+
+	bStartWithPingEnabled = true;
+}
+
+void USpatialPingComponent::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const
+{
+	Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+
+	DOREPLIFETIME(USpatialPingComponent, ReplicatedPingID);
+}
+
+void USpatialPingComponent::BeginPlay()
+{
+	Super::BeginPlay();
+	//Attempt to cast the owner of this component to the PlayerController class
+	//Component will do nothing if cast fails
+	OwningController = Cast<APlayerController>(GetOwner());
+	if (bStartWithPingEnabled)
+		SetPingEnabled(true);
+}
+
+void USpatialPingComponent::EndPlay(const EEndPlayReason::Type EndPlayReason)
+{
+	Super::EndPlay(EndPlayReason);
+	//Clear up the timer before the component is disabled or destroyed
+	SetPingEnabled(false);
+}
+
+bool USpatialPingComponent::GetIsPingEnabled() const
+{
+	return bIsPingEnabled;
+}
+
+void USpatialPingComponent::SetPingEnabled(bool bSetEnabled)
+{
+	if (OwningController) {
+		//Only execute on owning local client
+		if (OwningController->GetLocalRole() == ENetRole::ROLE_AutonomousProxy)
+		{
+			if (bSetEnabled && !bIsPingEnabled) {
+				EnablePing();
+			}
+			else if (!bSetEnabled && bIsPingEnabled) {
+				DisablePing();
+			}
+		}
+	}
+}
+
+float USpatialPingComponent::GetPing() const
+{
+	return RTPing;
+}
+
+void USpatialPingComponent::EnablePing()
+{
+	UWorld* World = GetWorld();
+	if (World)
+	{
+		//Set looping timer to 'tick' this component and send ping RPC with configurable frequency
+		World->GetTimerManager().SetTimer(PingTimerHandle, this, &USpatialPingComponent::TickPingComponent, PingFrequency, true);
+		bIsPingEnabled = true;
+	}
+}
+
+void USpatialPingComponent::DisablePing()
+{
+	UWorld* World = GetWorld();
+	if (World)
+	{
+		//Clear the timer
+		World->GetTimerManager().ClearTimer(PingTimerHandle);
+		bIsPingEnabled = false;
+		RTPing = 0.f;
+	}
+}
+
+void USpatialPingComponent::TickPingComponent()
+{
+	SendNewPing();
+	//Pass latest measured ping to owning controller to be processed by PlayerState
+	OwningController->UpdatePing(RTPing);
+}
+
+void USpatialPingComponent::SendNewPing()
+{
+	//Send a new ping using the current local time since start as both ID and timestamp
+	SendServerWorkerPingID(GetWorld()->GetRealTimeSeconds());
+}
+
+void USpatialPingComponent::OnRep_ReplicatedPingID()
+{
+	//If the new replicated ping timestamp is older than prev then we can ignore it
+	if (ReplicatedPingID < LastReceivedPingID)
```

Comment:
What situation can this happen in?

Reply:
I was under the impression that RPCs aren't necessarily guaranteed to arrive in the same order they were sent, but I believe I was mistaken so yeah... redundant check removed. It wouldn't really matter too much if they did arrive out of order anyway.

Reply:
Ah ok, good to know, thanks.